File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.py
from collections import deque
def solution(maps):
    answer = []
    n,m = len(maps), len(maps[0])
    dx = [0,0,-1,1]
    dy = [-1,1,0,0]
    
    def bfs(x,y,sm):
        cnt = sm
        q = deque()
        q.append((x,y))
        visited[x][y] = 1
        
        while q:
            x,y = q.popleft()
            
            for i in range(4):
                nx = x+dx[i]
                ny = y+dy[i]
                
                if 0<=nx<n and 0<=ny<m and not visited[nx][ny]:
                    if maps[nx][ny] !='X':
                        cnt+=int(maps[nx][ny])
                        visited[nx][ny]=1
                        q.append((nx,ny))
        return cnt
        
        
        
    visited =[[0]*m for _ in range(n)]
    for i in range(n):
        for j in range(m):
            if maps[i][j] !='X' and not visited[i][j]:
                visited[i][j] = 1
                answer.append(bfs(i,j,int(maps[i][j])))
                
                
    if len(answer) <1:
        answer.append(-1)
    else:
        answer.sort()
    return answer

File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.java
import java.util.*;
class Solution {
    static boolean[][] visited;
    static char[][] map;
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};
    public int[] solution(String[] maps) {
        List<Integer> answer = new ArrayList<>();
        int cnt =0;
        map = new char[maps.length][maps[0].length()];
        visited = new boolean[map.length][map[0].length];
        for(int i=0;i<maps.length;i++){
            map[i] = maps[i].toCharArray();
        }
        for(int i=0; i<map.length; i++){
            for(int j=0; j<map[0].length;j++){
                if(!visited[i][j] && map[i][j] != 'X'){
                    answer.add(bfs(i, j));
                }
            }
        }
        if(answer.size()==0){
            answer.add(-1);
        }
        Collections.sort(answer);
        
        int[] ans = answer.stream().mapToInt(Integer::intValue).toArray();
        
        return ans;
    }
    public static int bfs(int x, int y){
        int sum = 0;
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{x,y});
        visited[x][y] = true;
        
        while(!q.isEmpty()){
            int cur[] = q.poll();
            int cx = cur[0];
            int cy = cur[1];
            
            sum+= map[cx][cy] - '0';
            for(int i=0;i<4;i++){
                int nx = cx+dx[i];
                int ny = cy +dy[i];
                if(nx<0||ny<0||nx >=map.length|| ny>=map[0].length)
                    continue;
                if(!visited[nx][ny]&&map[nx][ny]!='X'){
                    visited[nx][ny] = true;
                    q.offer(new int[]{nx,ny});
                }
            }
        }
        return sum;
    }
}